### ISSUE TITLE
Regression in v4: expect(404) treats status code 404 as an error

### ISSUE THREAD

**Issue by bajtos** [visionmedia/superagent#1466]
Consider the following code snippet:

```js
const supertest = require('supertest');

supertest('http://example.com')
  .get('/path-not-found')
  .expect(404)
  .then(
    ok => console.log('PASS'),
    err => {
      console.error('FAIL', err.stack);
      process.exit(1);
    });
```

Run it using `supertest@3.4.2` to get the expected outcome:

```
PASS
```

Run it using `supertest@4.0.0` to reproduce the bug:

```
FAIL Error: Not Found
    at Test.Request.callback (/private/tmp/bug/node_modules/supertest/node_modules/superagent/lib/node/index.js:804:15)
    at IncomingMessage.parser (/private/tmp/bug/node_modules/supertest/node_modules/superagent/lib/node/index.js:1036:18)
    at Stream.emit (events.js:182:13)
    at Unzip.unzip.on (/private/tmp/bug/node_modules/supertest/node_modules/superagent/lib/node/unzip.js:55:12)
    at Unzip.emit (events.js:187:15)
    at endReadableNT (_stream_readable.js:1094:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

**Comment by bajtos** [comment]
I have reviewed the changes in supertest and superagent, but not was able to find the source of the problem.

- supertest changes from v3 to v4: https://github.com/visionmedia/supertest/compare/v3.4.2...v4.0.0
- superagent changes from v3 to v4: https://github.com/visionmedia/superagent/compare/v3.8.3...v4.1.0

**Comment by dejanr** [comment]
I noticed the same this morning, it must be something related to promises, as by using the callback it works.

**Comment by raymondfeng** [comment]
I can confirm the issue is introduced by superagent instead of supertest.

**Comment by raymondfeng** [comment]
The root cause seems to be https://github.com/visionmedia/superagent/compare/v3.8.3...v4.0.0#diff-be27821c599006faab0d29e2cace0392R238. Commenting it out fixes our issue.

**Comment by midrissi** [comment]
Actually, this is the commit that causes the issue: https://github.com/visionmedia/superagent/commit/6670a0fb708fef3d8e09538ee8af6c73f682f346#diff-be27821c599006faab0d29e2cace0392

**Comment by holm** [comment]
This was definitely surprising to us. It seems to be listed as a "minor change" int he history, but its not easy to understand what "all error events" mean.

I think this will be very surprising to most users, since there is a very big difference between catching an error and checking for as status.

**Comment by bajtos** [comment]
I am not able to reproduce the problem when calling `superagent` API directly. It looks like the `end` method provided by supertest ([source](https://github.com/visionmedia/supertest/blob/0cfc1983e3ffcbaf5360d0c29e9b6c4e0942a26b/lib/test.js#L120-L136)) is executed correctly, but only after the error detected via `this.on('error')` has already rejected the promise here:

https://github.com/visionmedia/superagent/blob/0f0949fa8a93accf788b874f2b785740de6d0918/lib/request-base.js#L245

The following change seems to fix the problem for me:

```diff
- self.on('error', innerReject);
+ self.on('error', () => { setImmediate(innerReject); });
```

However, I find it very hacky. It essentially lets the `end` callback to execute `innerResolve` first, and later lets `on('error')` callback to call `innerReject` which is ignored by the Promise runtime.

Personally, I think we should revert https://github.com/visionmedia/superagent/commit/6670a0fb708fef3d8e09538ee8af6c73f682f346#diff-be27821c599006faab0d29e2cace0392 and look for a more robust fix for #1370. One that fixes the problem both for `.end()` and `.then()` users.

At the moment, using `superagent@4.1.0`, the following code adopted for callbacks from #1370 is crashing the process on unhandled rejection:

```js
require('superagent').post('http://example.com/does-not-exist')
  .field({a:1,b:2})
  .attach('c', 'does-not-exists.txt')
  //.on('error', e => console.log('ERR', e.message))
  .end((err, res) => {
    if (err) console.log('CATCH', e.message);
    else console.log('THEN');
  });
```

Output:

```
events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open 'does-not-exists.txt'
Emitted 'error' event at:
    at FormData._formData.on.err (/private/tmp/node_modules/superagent/lib/node/index.js:267:12)
    at FormData.emit (events.js:187:15)
    at FormData.CombinedStream._emitError (/private/tmp/node_modules/combined-stream/lib/combined_stream.js:188:8)
    at DelayedStream.<anonymous> (/private/tmp/node_modules/combined-stream/lib/combined_stream.js:114:10)
    at DelayedStream.emit (events.js:187:15)
    at DelayedStream.<anonymous> (/private/tmp/node_modules/delayed-stream/lib/delayed_stream.js:69:15)
    at Array.forEach (<anonymous>)
    at DelayedStream.release (/private/tmp/node_modules/delayed-stream/lib/delayed_stream.js:68:24)
    at DelayedStream.resume (/private/tmp/node_modules/delayed-stream/lib/delayed_stream.js:55:10)
    at DelayedStream.pipe (/private/tmp/node_modules/delayed-stream/lib/delayed_stream.js:76:8)
```

**Comment by bajtos** [comment]
I think I have a reasonably fix for this problem, see https://github.com/visionmedia/superagent/pull/1468

**Comment by bajtos** [comment]
Closing as fixed by #1468
